titus-server-master/src/main/java/com/netflix/titus/master/scheduler/SchedulerUtils.java

```diff
@@ -52,4 +58,47 @@ public static String getAttributeValue(VirtualMachineCurrentState targetVM, Stri
         Protos.Attribute attribute = targetVM.getCurrAvailableResources().getAttributeMap().get(attributeName);
         return Strings.nullToEmpty(attribute.getText().getValue());
     }
+
+    public static String getZoneId(VirtualMachineCurrentState targetVM, String zoneAttributeName) {
+        Map<String, Protos.Attribute> attributeMap = targetVM.getCurrAvailableResources().getAttributeMap();
+        if (attributeMap == null) {
+            return null;
+        }
+
+        Protos.Attribute attributeValue = attributeMap.get(zoneAttributeName);
+        if (attributeValue == null) {
+            return null;
+        }
+
+        if (!attributeValue.hasText()) {
+            return null;
+        }
+        String value = attributeValue.getText().getValue();
+        return value == null || value.isEmpty() ? null : value;
+    }
+
+    public static Map<String, Integer> groupCurrentlyAssignedTasksByZoneId(String jobId, Collection<TaskTracker.ActiveTask> tasksCurrentlyAssigned, String zoneAttributeName) {
+        Map<String, Integer> result = new HashMap<>();
+
+        for (TaskTracker.ActiveTask activeTask : tasksCurrentlyAssigned) {
+            TaskRequest request = activeTask.getTaskRequest();
+            if (request instanceof V3QueueableTask) {
+                String requestJobId = ((V3QueueableTask) request).getJob().getId();
+                if (jobId.equals(requestJobId)) {
+                    Map<String, Protos.Attribute> attributeMap = activeTask.getTotalLease().getAttributeMap();
```

Comment:
Getting the value out of an attribute is duplicated again here

Reply:
Fixed